### PR TITLE
Source Instagram: fix bug with transient failures

### LINKED ISSUE #2626: Source instagram: retry transient server failures
## Tell us about the problem you're trying to solve
As demonstrated in https://github.com/airbytehq/airbyte/issues/1814#issuecomment-808337399 when the Instagram source receives a transient 500 from the FB servers, the sync for that stream fails. 

## Describe the solution you’d like
The source should instead backoff & retry on transient failures, same as with rate limiting. We should do this for all 500 errors.

### PR DESCRIPTION
**Pull request by yevhenii-ldv**
## What
* *Update backoff function for working with transient failures*
* *Add clear pagination for media resources to be sure that the backoff will always take place*

## Pre-merge Checklist
- [ ] *Run integration tests*
- [ ] *Publish Docker images*

## Recommended reading order
1. `test.java`
1. `component.ts`
1. the rest

### CODE REVIEW

File: airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py [hunk 1]
```diff
@@ -78,6 +79,22 @@ def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
...
+    @backoff_policy
+    def load_next_page(self, instance: Cursor):
+        instance.load_next_page()
+
+    @backoff_policy
+    def get_instance_cursor(self, ig_user: IGUser, attribute: str, params: dict = None, fields: Sequence[str] = None) -> Cursor:
+        return getattr(ig_user, attribute)(params=params, fields=fields)
+
+    def pagination(self, instance: Cursor) -> Iterator[Any]:
+        yield from instance._queue
```
Comment (arhip11) [inline airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py:91]: It looks like we can't decorate pagination with backoff without accessing private attributes of the Cursor.
Let's document this approach via comment and insert TODO warning about possible breaking changes in Facebook API implementation.
We can consider creating an issue for facebook_business client.

File: airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py [hunk 2]
```diff
@@ -272,13 +286,10 @@ def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
...
-    @backoff_policy
     def _get_media(self, instagram_user: IGUser, params: dict = None, fields: Sequence[str] = None) -> Iterator[Any]:
-        """
-        This is necessary because the functions that call this endpoint return
-        a generator, whose calls need decorated with a backoff.
-        """
-        return instagram_user.get_media(params=params, fields=fields)
+        media = self.get_instance_cursor(instagram_user, "get_media", params=params, fields=fields)
+        for media_obj in self.pagination(media):
+            yield media_obj
```
Comment (arhip11) [inline airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py:292]: `_get_media` and `_get_stories` looks identical except "get_*" attribute name string.
We can consider moving `get_instance_cursor` into `pagination` and extend `pagination` parameters, e. g.:
```
for media_obj in self.pagination("get_media"):
    yield media_obj
```
Or we can consider having wrapper function for the 3 lines above.

File: airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py [hunk 3]
```diff
@@ -301,13 +312,10 @@ def list(self, fields: Sequence[str] = None) -> Iterator[dict]:
...
-    @backoff_policy
     def _get_stories(self, instagram_user: IGUser, params: dict, fields: Sequence[str] = None) -> Iterator[Any]:
-        """
-        This is necessary because the functions that call this endpoint return
-        a generator, whose calls need decorated with a backoff.
-        """
-        return instagram_user.get_stories(params=params, fields=fields)
+        stories = self.get_instance_cursor(instagram_user, "get_stories", params=params, fields=fields)
+        for story_obj in self.pagination(stories):
+            yield story_obj
```
Comment (arhip11) [inline airbyte-integrations/connectors/source-instagram/source_instagram/client/api.py:318]: Please, see comment regarding `_get_media`.

**Review by arhip11 — CHANGES_REQUESTED** [review]

**Review by arhip11 — APPROVED** [review]

File: airbyte-integrations/connectors/source-instagram/source_instagram/client/common.py [hunk 4]
```diff
@@ -40,11 +39,7 @@ def log_retry_attempt(details):
         logger.info(f"Caught retryable error after {details['tries']} tries. Waiting {details['wait']} more seconds then retrying...")
 
     def should_retry_api_error(exc):
-        if exc.http_status() == status_codes.TOO_MANY_REQUESTS or (
-            exc.http_status() == status_codes.FORBIDDEN and exc.api_error_message() == "(#4) Application request limit reached"
-        ):
-            return True
-        return False
+        return exc.api_transient_error()
```
Comment (sherifnada) [inline airbyte-integrations/connectors/source-instagram/source_instagram/client/common.py:42]: are you sure that this also applies to rate limiting? I don't see it in docs anywhere. Why not just add this as an OR clause?

**Review by sherifnada — CHANGES_REQUESTED** [review]
Request changes because of worry about a regression

File: airbyte-integrations/connectors/source-instagram/source_instagram/client/common.py (hunk 4 above)
Comment (yevhenii-ldv) [reply airbyte-integrations/connectors/source-instagram/source_instagram/client/common.py]: I found information that this value also applies to Rate Limiting, but now I could not find it. 
To be more safely, I added this check to the general list of errors to be handled using `backoff`.

**Review by sherifnada — APPROVED** [review]

File: airbyte-integrations/connectors/source-instagram/source_instagram/client/common.py (hunk 4 above)
Comment (sherifnada) [reply airbyte-integrations/connectors/source-instagram/source_instagram/client/common.py]: SGTM

**Pull request merged**